contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -11,66 +11,150 @@ contract KeystoneForwarder is IForwarder, ConfirmedOwner, TypeAndVersionInterfac
   error ReentrantCall();
 
   /// @notice This error is returned when the data with report is invalid.
-  /// This can happen if the data is shorter than SELECTOR_LENGTH + REPORT_LENGTH.
+  /// This can happen if the data is shorter than SELECTOR_LENGTH + REPORT_HEADER_LENGTH.
   /// @param data the data that was received
   error InvalidData(bytes data);
 
+  /// @notice This error is thrown whenever trying to set a config
+  /// with a fault tolerance of 0
+  error FaultToleranceMustBePositive();
+
+  /// @notice This error is thrown whenever a report is signed
+  /// with more than the max number of signers
+  /// @param numSigners The number of signers who have signed the report
+  /// @param maxSigners The maximum number of signers that can sign a report
+  error ExcessSigners(uint256 numSigners, uint256 maxSigners);
+
+  /// @notice This error is thrown whenever a report is signed
+  /// with less than the minimum number of signers
+  /// @param numSigners The number of signers who have signed the report
+  /// @param minSigners The minimum number of signers that need to sign a report
+  error InsufficientSigners(uint256 numSigners, uint256 minSigners);
+
+  error RepeatedSigner();
+
+  error WrongNumberOfSignatures();
+
+  error InvalidDonId();
+  error InvalidSigner();
+
   uint256 private constant SELECTOR_LENGTH = 4;
-  uint256 private constant REPORT_LENGTH = 64;
 
-  struct HotVars {
-    bool reentrancyGuard; // guard against reentrancy
-  }
+  bool internal s_reentrancyGuard; // guard against reentrancy
 
-  HotVars internal s_hotVars; // Mixture of config and state, commonly accessed
+  /// @notice Contains the signing address of each oracle
+  struct OracleSet {
+    uint8 f; // Number of faulty nodes allowed
+    address[] signers;
+    mapping(address => uint256) _positions; // 1-indexed to detect unset values
+  }
+  mapping(bytes4 donId => OracleSet) internal s_configs;
 
-  mapping(bytes32 => address) internal s_reports;
+  // reportId = keccak256(bytes20(receiver) | selector | workflowExecutionId)
+  mapping(bytes32 reportId => address transmitter) internal s_reports;
 
   constructor() ConfirmedOwner(msg.sender) {}
 
-  // send a report to targetAddress
+  uint256 internal constant MAX_ORACLES = 31;
+
+  // NOTE: we don't inherit OCR2Base since unlike aggregator we only care about signers, not transmitters
+  // and the signers don't fetch their config from the forwarder
```

Comment:
OCR2Base is not meant to be inherited in prod (https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/shared/ocr2/OCR2Base.sol#L8-L14) so this comment is redundant.

Reply:
Fun fact: Functions inherits this in prod ;) (it was obviously included in the audit).  I agree with comment redundancy though.